README.md

```diff
@@ -1,133 +1,616 @@
+├── package.json
+├── src
+│   ├── Weather.ts
+│   ├── WeatherClient.ts
+│   ├── commands
+│   ├── extensionConfiguration.ts
+│   ├── index.ts
+│   ├── models
+│   ├── pagination
+│   ├── runtimeConfig.browser.ts
+│   ├── runtimeConfig.native.ts
+│   ├── runtimeConfig.shared.ts
+│   ├── runtimeConfig.ts
+│   └── runtimeExtensions.ts
+├── tsconfig.cjs.json
+├── tsconfig.es.json
+├── tsconfig.json
+├── tsconfig.types.json
+└── typedoc.json
+```
+
+#### Code generation implementations not included
+
+Smithy TypeScript provides default code generation implementations for generating TypeScript clients, but also requires customers to either implement or consume certain implementations where there is no default.
+
+For Smithy TypeScript clients, the main implementations not provided are protocol generators and handling endpoint resolution (see [the TypeScript code generation section](#typescript-code-generation)).
+
+> If there are items missing from this section, feel free to [create an issue](https://github.com/awslabs/smithy-typescript/issues/new).
+
+##### Protocols
+
+Protocols define how operation inputs and outputs are serialized and deserialized on the wire. This behavior can be defined in Smithy through [protocol traits](https://smithy.io/2.0/spec/protocol-traits.html) with corresponding implementations of [the `ProtocolGenerator` interface](smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/ProtocolGenerator.java) in Smithy TypeScript. Besides the `ProtocolGenerator` interface, Smithy TypeScript has additional abstract classes that partially implement the `ProtocolGenerator` interface and can be extended: [`HttpBindingProtocolGenerator` for HTTP binding protocols](smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java) and [`HttpRpcProtocolGenerator` for HTTP RPC protocols](smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpRpcProtocolGenerator.java).
+
+Once a `ProtocolGenerator` is implemented, the implementation can be registered through a `TypeScriptIntegration`:
+
+- `TypeScriptIntegration` with `ProtocolGenerator` implementation:
+  ```java
+  // src/main/java/typescript/example/client/gradle/ExampleClientProtocolGeneratorIntegration.java
+  package typescript.example.client.gradle;
+  
+  // ...
+  
+  public class ExampleClientProtocolGeneratorIntegration implements TypeScriptIntegration {
+      // ProtocolGenerator implementation is inline for brevity, but should be in its own file
+      private static class ExampleClientProtocolGenerator implements ProtocolGenerator {
+          // Implement ProtocolGenerator methods
+      }
+
+      @Override
+      public List<ProtocolGenerator> getProtocolGenerators() {
+          return List.of(new ExampleClientProtocolGenerator());
+      }
+  }
+  ```
+- Registering the `TypeScriptIntegration`:
```

Comment:
We should probably call out the use of the Java SPI here explicitly. Ive found that people can be confused how integrations are pulled in otherwise.

Reply:
Both sections mentioning `TypeScriptIntegration` point to the "Customizations via `TypeScriptIntegration`" which talks about Java SPI.

Maybe this is an assumption, but I'm not sure people are that interested in the Java intricacies.